### ISSUE TITLE
The implementation of CustomRun is not aligned with what we documented.

### ISSUE THREAD

**Issue by XinruZhang** [tektoncd/pipeline#5582]
# Expected Behavior

We documented that `timeout` in CustomRun is for constrainting the running time of all retry attempts, see [TEP-069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md). If we implement a CustomRun controller as the documentation guided, and use it in a Pipeline:
```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: custom-task-pipeline
spec:
  tasks:
  - name: wait
    timeout: "1s"
    retries: 2
    taskRef:
      apiVersion: wait.testing.tekton.dev/v1alpha1
      kind: Wait
    params:
    - name: duration
      value: "2s"

---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: pr-custom-task-
spec:
  pipelineRef:
    apiVersion: tekton.dev/v1beta1
    name: custom-task-pipeline
```

In Custom Controller, it detects that the run has timedout, so marked it as timedout, won't update `retriesStatus`.

And we expect the PipelineRun failed right after the CustomRun been marked as failed on timedout.

# Actual Behavior

The PipelineRun never treats the CustomRun as `timedout`, instead, it treat the CustomRun as `Running` forever until PipelineRun iteself hits its Timeout limit (`default.DefaultTimeoutMinutes` if the `timeout` is not set). Because the `len(run.retriesStatus)` is always less than `retries`.

# Steps to Reproduce the Problem

1. Have a Custom Task controller:
    - Can run a specific period of time.
    - Support retries and timeout as documented.
3. Submit a PipelineRun that contains the CustomTask.
    - Set the `timeout` less than the running time of the Run.
    - Set `retriesStatus` > 0

The PipelineRun will fail after `default.DefaultTimeoutMinutes`.

**Comment by XinruZhang** [comment]
The current [implementation](https://github.com/tektoncd/pipeline/blob/e1fd72720652fa5a1b7f811a6ce1d6c943be11a6/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L198) in PipelineRun reconciler, it [relies on both Condition and retriesStatus](https://github.com/tektoncd/pipeline/blob/e1fd72720652fa5a1b7f811a6ce1d6c943be11a6/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L198): `run.IsDone() & len(retriesStatus)==retries`
```
// isFailure returns true only if the run has failed and will not be retried.
// If the PipelineTask has a Matrix, isFailure returns true if any run has failed (no remaining retries)
// and all other runs are done.
func (t ResolvedPipelineTask) isFailure() bool {
	if t.isCancelledForTimeOut() {
		return true
	}
	if t.isCancelled() {
		return true
	}
	if t.isSuccessful() {
		return false
	}
	var c *apis.Condition
	var isDone bool

	switch {
	case t.IsCustomTask() && t.IsMatrixed():
		if len(t.Runs) == 0 {
			return false
		}
		isDone = true
		atLeastOneFailed := false
		for _, run := range t.Runs {
			isDone = isDone && run.IsDone()
			runFailed := run.Status.GetCondition(apis.ConditionSucceeded).IsFalse() && !t.hasRemainingRetries()
			atLeastOneFailed = atLeastOneFailed || runFailed
		}
		return atLeastOneFailed && isDone
	case t.IsCustomTask():
		if t.Run == nil {
			return false
		}
		c = t.Run.Status.GetCondition(apis.ConditionSucceeded)
		isDone = t.Run.IsDone()
	case t.IsMatrixed():
		if len(t.TaskRuns) == 0 {
			return false
		}
		isDone = true
		atLeastOneFailed := false
		for _, taskRun := range t.TaskRuns {
			isDone = isDone && taskRun.IsDone()
			taskRunFailed := taskRun.Status.GetCondition(apis.ConditionSucceeded).IsFalse() && !t.hasRemainingRetries()
			atLeastOneFailed = atLeastOneFailed || taskRunFailed
		}
		return atLeastOneFailed && isDone
	default:
		if t.TaskRun == nil {
			return false
		}
		c = t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
		isDone = t.TaskRun.IsDone()
	}
	return isDone && c.IsFalse() && !t.hasRemainingRetries()
}
```

**Comment by ScrapCodes** [comment]
As per the developer Guide, https://tekton.dev/docs/pipelines/runs/#developer-guide-for-custom-controllers-supporting-timeout

A Custom task controller has to perform : 
```go
			run.Status.MarkRunFailed(reason, "Run %s/%s was cancelled due to Timeout.", run.Namespace, run.Name)

```
When it observes, ` run.IsCancelled()  ` as true.

This should make `run.IsDone()` true and `run.IsFailed()` evaluate true as well.

This is the expectation, what do you observe in your code? Can you show us what you are running? or may be give an example reproducer?

**Comment by XinruZhang** [comment]
Yeah, `run.IsDone()` and `run.IsFailed()` will be true, but `!t.hasRemainingRetries()` will be false. So this run will be never considered as Done in `Resolved.PipelineTask.IsDone()` https://github.com/tektoncd/pipeline/blob/e1fd72720652fa5a1b7f811a6ce1d6c943be11a6/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L76

Therefore, it will be always treated as schedulable, but when the PipelineRun controller wants to schedule it (create a new run), it will find that the run already exists.

I created a [custom run controller](https://github.com/XinruZhang/xinru-custom-task) that implements the functionality as described in the developer guide. Install it locally, then apply the following yamls, you'll reproduce the error.

```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: custom-task-pipeline
spec:
  tasks:
  - name: wait
    timeout: "1s"
    retries: 1
    taskRef:
      apiVersion: wait.testing.tekton.dev/v1alpha1
      kind: Wait
    params:
    - name: duration
      value: "2s"
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: pr-custom-task-
spec:
  pipelineRef:
    apiVersion: tekton.dev/v1beta1
    name: custom-task-pipeline
```

See the result:
![image](https://user-images.githubusercontent.com/40881156/193304671-d650c81b-5a3b-4bca-8b72-25d072a1c3ce.png)


```
❯ kubectl get run pr-custom-task-gdpqw-wait -o yaml
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1beta1","kind":"Pipeline","metadata":{"annotations":{},"name":"custom-task-pipeline","namespace":"default"},"spec":{"tasks":[{"name":"wait","params":[{"name":"duration","value":"2s"}],"retries":1,"taskRef":{"apiVersion":"wait.testing.tekton.dev/v1alpha1","kind":"Wait"},"timeout":"1s"}]}}
  creationTimestamp: "2022-09-30T15:28:41Z"
  generation: 1
  labels:
    tekton.dev/memberOf: tasks
    tekton.dev/pipeline: custom-task-pipeline
    tekton.dev/pipelineRun: pr-custom-task-gdpqw
    tekton.dev/pipelineTask: wait
  name: pr-custom-task-gdpqw-wait
  namespace: default
  ownerReferences:
  - apiVersion: tekton.dev/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: PipelineRun
    name: pr-custom-task-gdpqw
    uid: 3cb51ad5-6a19-4355-819e-52cba778198d
  resourceVersion: "53449752"
  uid: 357684a8-8b90-4532-bc2d-620ba0564e26
spec:
  params:
  - name: duration
    value: 2s
  ref:
    apiVersion: wait.testing.tekton.dev/v1alpha1
    kind: Wait
  retries: 1
  serviceAccountName: default
  timeout: 1s
status:
  completionTime: "2022-09-30T15:28:42Z"
  conditions:
  - lastTransitionTime: "2022-09-30T15:28:42Z"
    message: Custom Task Run %!v(MISSING) timed out
    reason: TimedOut
    status: "False"
    type: Succeeded
  extraFields: null
  observedGeneration: 1
  startTime: "2022-09-30T15:28:41Z"
```
```
❯ kubectl get pr pr-custom-task-gdpqw -o yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"tekton.dev/v1beta1","kind":"Pipeline","metadata":{"annotations":{},"name":"custom-task-pipeline","namespace":"default"},"spec":{"tasks":[{"name":"wait","params":[{"name":"duration","value":"2s"}],"retries":1,"taskRef":{"apiVersion":"wait.testing.tekton.dev/v1alpha1","kind":"Wait"},"timeout":"1s"}]}}
  creationTimestamp: "2022-09-30T15:28:41Z"
  generateName: pr-custom-task-
  generation: 1
  labels:
    tekton.dev/pipeline: custom-task-pipeline
  name: pr-custom-task-gdpqw
  namespace: default
  resourceVersion: "53449738"
  uid: 3cb51ad5-6a19-4355-819e-52cba778198d
spec:
  pipelineRef:
    apiVersion: tekton.dev/v1beta1
    name: custom-task-pipeline
  serviceAccountName: default
  timeout: 1h0m0s
status:
  childReferences:
  - apiVersion: tekton.dev/v1alpha1
    kind: Run
    name: pr-custom-task-gdpqw-wait
    pipelineTaskName: wait
  conditions:
  - lastTransitionTime: "2022-09-30T15:28:41Z"
    message: 'Tasks Completed: 0 (Failed: 0, Cancelled 0), Incomplete: 1, Skipped:
      0'
    reason: Running
    status: Unknown
    type: Succeeded
  pipelineSpec:
    tasks:
    - name: wait
      params:
      - name: duration
        value: 2s
      retries: 1
      taskRef:
        apiVersion: wait.testing.tekton.dev/v1alpha1
        kind: Wait
      timeout: 1s
  startTime: "2022-09-30T15:28:41Z"
```

```
❯ kubectl describe pr pr-custom-task-gdpqw
Name:         pr-custom-task-gdpqw
Namespace:    default
Labels:       tekton.dev/pipeline=custom-task-pipeline
Annotations:  <none>
API Version:  tekton.dev/v1beta1
Kind:         PipelineRun
Metadata:
  Creation Timestamp:  2022-09-30T15:28:41Z
  Generate Name:       pr-custom-task-
  Generation:          1
  Managed Fields:
    API Version:  tekton.dev/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
        f:labels:
          .:
          f:tekton.dev/pipeline:
    Manager:      controller
    Operation:    Update
    Time:         2022-09-30T15:28:41Z
    API Version:  tekton.dev/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:childReferences:
        f:conditions:
        f:pipelineSpec:
          .:
          f:tasks:
        f:startTime:
    Manager:      controller
    Operation:    Update
    Subresource:  status
    Time:         2022-09-30T15:28:41Z
    API Version:  tekton.dev/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName:
      f:spec:
        .:
        f:pipelineRef:
          .:
          f:apiVersion:
          f:name:
    Manager:         kubectl-create
    Operation:       Update
    Time:            2022-09-30T15:28:41Z
  Resource Version:  53449738
  UID:               3cb51ad5-6a19-4355-819e-52cba778198d
Spec:
  Pipeline Ref:
    API Version:         tekton.dev/v1beta1
    Name:                custom-task-pipeline
  Service Account Name:  default
  Timeout:               1h0m0s
Status:
  Child References:
    API Version:         tekton.dev/v1alpha1
    Kind:                Run
    Name:                pr-custom-task-gdpqw-wait
    Pipeline Task Name:  wait
  Conditions:
    Last Transition Time:  2022-09-30T15:28:41Z
    Message:               Tasks Completed: 0 (Failed: 0, Cancelled 0), Incomplete: 1, Skipped: 0
    Reason:                Running
    Status:                Unknown
    Type:                  Succeeded
  Pipeline Spec:
    Tasks:
      Name:  wait
      Params:
        Name:   duration
        Value:  2s
      Retries:  1
      Task Ref:
        API Version:  wait.testing.tekton.dev/v1alpha1
        Kind:         Wait
      Timeout:        1s
  Start Time:         2022-09-30T15:28:41Z
Events:
  Type     Reason         Age                 From         Message
  ----     ------         ----                ----         -------
  Normal   Started        16m                 PipelineRun
  Normal   Running        16m                 PipelineRun  Tasks Completed: 0 (Failed: 0, Cancelled 0), Incomplete: 1, Skipped: 0
  Warning  InternalError  16m (x11 over 16m)  PipelineRun  1 error occurred:
           * error creating Run called pr-custom-task-gdpqw-wait for PipelineTask wait from PipelineRun pr-custom-task-gdpqw: runs.tekton.dev "pr-custom-task-gdpqw-wait" already exists
  Warning  RunCreationFailed  5m19s (x18 over 16m)  PipelineRun  Failed to create Run "pr-custom-task-gdpqw-wait": runs.tekton.dev "pr-custom-task-gdpqw-wait" already exists
```

**Comment by jerop** [comment]
This issue was fixed in v0.43 through implementation of [TEP-0121](https://github.com/tektoncd/community/blob/main/teps/0121-refine-retries-for-taskruns-and-customruns.md) and [TEP-0114](https://github.com/tektoncd/community/blob/main/teps/0114-custom-tasks-beta.md)

/close
